block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockMetadataStore.scala

```diff
@@ -64,8 +76,28 @@ object BlockMetadataStore {
         _ <- store.put(block.blockHash, block)
       } yield ()
 
-    def addFinalizedBlock(finalizedBlockHash: BlockHash): F[Unit] =
-      dagState.modify(st => st.copy(finalizedBlockSet = st.finalizedBlockSet + finalizedBlockHash))
+    def recordFinalized(hashes: List[BlockHash])(implicit sync: Sync[F]): F[Unit] =
+      for {
+        _ <- dagState.modify { st =>
+              st.copy(finalizedBlockSet = st.finalizedBlockSet ++ hashes)
+            }
+        curVs <- store.getUnsafe(hashes)
+        newVs = curVs.map(_.copy(finalized = true))
+        _     <- store.put(newVs.map(v => (v.blockHash, v)))
+      } yield ()
+
+    def recordDirectlyFinalised(hash: BlockHash)(implicit sync: Sync[F]): F[Unit] =
+      for {
+        curMeta <- store.getUnsafe(hash)
+        _       <- store.put(hash, curMeta.copy(directlyFinalized = true)) // update persistent meta
+        _ <- dagState.modify(
+              st =>
+                if (st.lastFinalizedBlock.exists { case (_, height) => height > curMeta.blockNum })
+                  st
+                else // update inmem state if new LFB higher then existing
+                  st.copy(lastFinalizedBlock = (hash, curMeta.blockNum).some)
+            )
+      } yield ()
```

Comment:
It would be nice to combine these two functions and have atomic update of all finalized blocks. :thinking:

Reply:
Now I'm thinking that its quite odd that we run finalization on DagRepresentation, and then get back to DagStorage to update it, and DagStorage actually might have been already updated in between.

Reply:
I'd better not touch it now, its quite a rabbit whole